src/definitions/x509/crl.rs

```diff
@@ -0,0 +1,297 @@
+use asn1_rs::{FromDer, Sequence};
+use const_oid::{AssociatedOid, ObjectIdentifier};
+use der::{Any, Decode, SliceReader};
+use signature::Verifier;
+use x509_cert::{
+    crl::{CertificateList, RevokedCert, TbsCertList},
+    ext::pkix::{
+        name::{DistributionPointName, GeneralName},
+        CrlDistributionPoints, CrlReason,
+    },
+    spki::AlgorithmIdentifierOwned,
+    TbsCertificate,
+};
+
+#[derive(Debug)]
+pub enum Error {
+    CertRevoked(Option<CrlReason>),
+
+    UnableToParseCrlExtension(der::Error),
+    UnableToParseCrlCertList(der::Error),
+    UnableToParseCurveKindExt(der::Error),
+    DistributionPointMalformed(&'static str),
+    FetchingCrl(reqwest::Error),
+    UnableToReachDistributionPoint,
+    IssuerMismatchBetweenCertAndCrl,
+    UnknownSignatureAlgorithm(Box<AlgorithmIdentifierOwned>),
+    SignatureTypeMismatch(Box<AlgorithmIdentifierOwned>, Box<AlgorithmIdentifierOwned>),
+    ExtractCrlSequenceSignature(asn1_rs::Err<asn1_rs::Error>),
+    UnableToParseCrlReason(der::Error),
+    MissingCurve,
+    UnknownCurveExt(ObjectIdentifier),
+    MissingPublicKey,
+    SignatureCheckOfCrlFailed,
+    InvalidPublicKeyFormat,
+    MissingSignature,
+    SignatureInWrongFormat,
+}
+
+pub const OID_EC_CURVE_P256: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.10045.3.1.7");
+pub const OID_EC_CURVE_P384: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.3.132.0.34");
+pub const OID_PUBLIC_KEY_ELLIPTIC_CURVE: ObjectIdentifier =
+    ObjectIdentifier::new_unwrap("1.2.840.10045.2.1");
+
+const OID_EXTENSION_REASON_CODE: ObjectIdentifier = ObjectIdentifier::new_unwrap("2.5.29.21");
+
+/// Given a cert, download and verify the associated crl listed in the cert, and verify the cert
+/// against the crl
+pub async fn fetch_and_validate_crl(cert: &TbsCertificate) -> Result<(), Error> {
```

Comment:
In general the CRL isn't necessarily signed by the PK in the cert. For 18013-5 the IACA root cert is the signer for all CRLs in the certificate chain, so you might need to update this function to take in a "crl_signer_cert" parameter.

Reply:
So I tried a slightly different strategy since there are other certs that need to be checked, but we don't want to have to download the CRL multiple times.

I therefore separated fetch_and_validate_crl from check_cert_against cert_lists, allowing higher level code with more context of the cert tree to iterate on that structure, fetching CRL and validating it against the root cert, then separately using that CRL against any other certs it sees fit.

Does that work for you?

Reply:
I'm not sure this is quite correct. From my understanding it is possible for a child certificate to have a different CRL endpoint than the root. Perhaps I'm wrong on that?

Reply:
@justAnIdentity do you know?

Reply:
I've changed "root_cert" to "crl_signing_cert" to hopefully alleviate these concerns.